pkg/netpol/eval/resources.go

```diff
@@ -574,48 +585,76 @@ func (pe *PolicyEngine) GetSelectedPeers(selectors labels.Selector, namespace st
-	if objLabels != nil {
-		nsLabels = objLabels.NsLabels
-		podLabels = objLabels.PodLabels
-	}
-
+// AddPodByNameAndNamespace adds a new fake pod to the pe.podsMap, used for adding ingress-controller pod
+func (pe *PolicyEngine) AddPodByNameAndNamespace(name, ns string) (Peer, error) {
 	podStr := types.NamespacedName{Namespace: ns, Name: name}.String()
 	newPod := &k8s.Pod{
 		Name:      name,
 		Namespace: ns,
 		FakePod:   true,
-		Labels:    podLabels,
 	}
-	if err := pe.resolveSingleMissingNamespace(ns, nsLabels); err != nil {
+	if err := pe.resolveSingleMissingNamespace(ns); err != nil {
 		return nil, err
 	}
-	// if exposure-analysis and this is not a fake ingress-controller
-	if pe.exposureAnalysisFlag && newPod.IsPodRepresentative() {
-		// first compute a unique string from labels to be used as a map key
-		keyStrFromLabels := k8s.VariantFromLabelsMap(objLabels.NsLabels) + "/" + k8s.VariantFromLabelsMap(objLabels.PodLabels)
-		if _, ok := pe.representativePeersMap[keyStrFromLabels]; ok { // we already have a representative peer with same labels
-			return nil, nil
-		}
-		newRepresentativePeer := &k8s.RepresentativePeer{Pod: newPod, PotentialNamespaceLabels: nsLabels}
-		pe.representativePeersMap[keyStrFromLabels] = newRepresentativePeer
-		return newRepresentativePeer, nil
-	}
-	// ingress-controller will be treated as a real pod, may be added to podsMap
 	pe.podsMap[podStr] = newPod
 	return &k8s.WorkloadPeer{Pod: newPod}, nil
 }
+
+// addRepresentativePod adds a new representative pod to the policy-engine (to pe.representativePeersMap).
+// if the given namespace string (podNs) is not empty (i.e. a real (policy's) namespace name), it will be assigned to the pod's Namespace;
+// and the "namespace name" requirement of the representative pod will be stored in its RepresentativeNsLabelSelector field.
+// if the representative pod will have no namespace (will not add a representative namespace to the policy-engine).
+// this func is used only with exposure-analysis
+func (pe *PolicyEngine) addRepresentativePod(podNs string, objSelectors *k8s.SingleRuleSelectors) error {
+	if objSelectors == nil { // should not get here
+		return errors.New(netpolerrors.NilRepresentativePodSelectorsErr)
+	}
+	nsLabelSelector := objSelectors.NsSelector
+	if nsLabelSelector == nil { // equal to podNs != ""
+		// if the objSelectors.NsSelector is nil, means inferred from a rule with nil nsSelector, which means the namespace of the
+		// pod is the namespace of the policy, so adding it as its RepresentativeNsLabelSelector requirement.
+		// by this, we ensure a representative peer may only represent the rule it was inferred from
+		// and uniqueness of representative peers.
+		// (another policy in another namespace, may have a rule with same podSelector, but the namespace will be different-
+		// so a different representative peer will be generated)
+		nsLabelSelector = &metav1.LabelSelector{MatchLabels: defaultNamespaceLabelsMap(podNs)}
```

Comment:
verify here that `podNs` is not empty

Reply:
meant to the branch `if nsLabelSelector == nil `

Reply:
done